### ISSUE TITLE
npm install doesn't include symlinks in a git repository

### ISSUE THREAD

**Issue by CodeLenny** [npm/npm#13050]
I've got a git repository I want to include as a package dependency, that includes several symlinks in it's git repository, that are necessary for the code to work.  When I used `npm install` to download it, the symlinks aren't included in `node_modules`.

Are symlinks supposed to be included during `npm install`?  I couldn't find any documentation suggesting that they weren't.

I recreated a minimal scenario with separate repos.  One ([npm-dep](https://gitlab.com/ryan.leonard-rweda/npm-dep)) includes a file, `a`, as well as a symlink to `a`, named `b`.

Using a separate repo ([npm-req](https://gitlab.com/ryan.leonard-rweda/npm-req)), I `npm install`ed the repository both via `git+ssh://` and `gitlab:`.  In each case, the symbolic link (`b`) wasn't included under `node_modules/npm-dep/`.

```sh
git clone git@gitlab.com:ryan.leonard-rweda/npm-dep.git
cd npm-req
npm init -y; git add package.json; git commit -m "package.json"
touch a
git add a; git commit -m "Added simple file 'a'"
ln -s a b
git add b; git commit -m "Added symlink 'b'"
git push -u origin master
cd ../
```

```sh
git clone git@gitlab.com:ryan.leonard-rweda/npm-req.git
cd npm-dep
npm init -y; git add package.json; git commit -m "package.json"
npm install --save git+ssh://git@gitlab.com:ryan.leonard-rweda/npm-dep.git
git commit -am "Added dependency."; git push -u origin master

ls node_modules/npm-dep/ # => a package.json
```

**Comment by piranna** [comment]
I've found symlinks are not included on packages published to npm registry, seems to be they are not included in the `npm pack` step.

**Comment by othiym23** [comment]
Symlinks within the package [aren't included](https://github.com/npm/fstream-npm/commit/71fd6255028168dde9fc60c073339481926d1e21#diff-bcfc576a0e4d3baa2f2454b33ace5868R155) in the packed tarballs, and this has been true since 2012 at least. @isaacs would have to weigh in on why, but if I were pressed to make a guess, I'd say that it's likely because there are situations in which links are either dangerous (due to [pointing at sensitive files outside the package root](https://github.com/npm/node-tar/issues/54#issuecomment-91067363) – see @isaacs's follow-up, though) or cause package unpacking to fail (shared folders under Vagrant, other filesystems where symlinks can't be created).

The latter makes it pretty tough to argue in favor of changing this behavior, because people developing under Vagrant is pretty common, and we get a lot of support traffic about bin-link creation failing (therefore requiring `--no-bin-links` for install). If you need links in order for a package to work, creating them via an `"install"` lifecycle script seems like the safest way to proceed, but it won't work for this group of users.

You're right that this is poorly documented, as are many of the rules governing what gets included, and what gets ignored, when creating package tarballs. I'm in the process of reworking a lot of that code due to #5082, and will probably be producing clearer developer documentation of what's included and what's not. If someone were to write a patch for `doc/cli/npm-pack.md` that documented some of these more obscure rules (or maybe just this one), I'd be happy to get it merged in. Thanks for your time! I hope this explanation is helpful, even if the behavior isn't!

**Comment by CodeLenny** [comment]
@othiym23 Thanks for providing input.  Enhanced documentation would certainly be good for the future.  I've created a little [package/unpackage module](https://gitlab.com/rweda/npm-pkg-sym) to run on `preversion`/`install`, that I'm including in packages that absolutely require symlinks, dropping support for the use-cases you outlined.  Thanks!

**Comment by isaacs** [comment]
Digging through my various archives mental and digital, the story seems to be that (at least in 2012) there was no way to make packing and unpacking symlinks work consistently or properly across Unix and Windows systems (and with mounts across unix and windows virtual machines, it isn't even trivial to detect whether it'll work ahead of time.)

On windows, you can create a junction to a file or directory without admin privs if you specify whether it's a dir or file (because they use different syscalls) and also if the target path is absolute.  However, if a file within a package contains a symbolic link from `../foo` to `./bar`, then it's not necessarily clear at unpack time whether foo is a file or directory, and making the link path absolute can create portability issues.  (You can no longer just pick up the folder and move it, for example, without also updating all the links.)

npm absolutely depends on symbolic links in some cases, and `npm link` should work on Windows as well as it does on Unix.  However, those are cases where the file/dir type and the absolute target path are known, and are managed programmatically, so not too hard to keep track of.  npm doesn't do much with package contents after unpacking in a target dir, so adding that level of bookkeeping would be really complicated.

Somewhat more prescriptively, relying on inter-package symbolic links makes code harder to reason about and read.  npm is generally pretty hands-off about letting you do you, but if you're reading someone else's code and you see `require('./foo')` but then you open up the file, and it's actually a symbolic link somewhere else, and does `require('./blarg')` and that's yet another symlink, it can trigger an unwanted confusion/frustration reaction and that's not great.  Reducing friction, that's the goal.  (If this was merely a ridiculous hypothetical, it'd be less compelling, but it was a real situation in a real production app that made Ryan and me both kinda table-flip and agree to disable this "feature".)

Lastly, there's security, which was really the nail in the coffin about why we opted to not even try to make this work.  If a package can include a symlink to `/etc/passwd` or some other thing, that's bad.  Some exploit poc's were developed, and it was clear that at least, npm had to do some careful analysis of any symlinks in packages that it installs, or put users at a very surprising risk.  Of course, if you're using a module, it can just have some malicious code that creates the symlink _anyway_, but at least we're not helping them.  Less likely than an exploit, per se, would be some very well-intended cleverness to say, "Gee, it's annoying that I have to keep requiring this other module I depend on by its path, I'll just symlink it up to the top level..." and then we're back at the rats' nest rage.

In summary, it was more work to do safely, in order to enable a feature which, while no doubt useful _sometimes_, is also a good-intentions-paved path to hell.  Doing extra work and adding more complexity to enable a thing that's almost always a bad idea seemed like a poor ROI.  Just preventing symbolic links in packages makes things a lot simpler, reduces windows headaches, improves security, and steers users towards alternative ways to solve their problems that don't cause headaches later.
